src/fastaq_handler.cpp

```diff
@@ -1,176 +1,77 @@
-            // skip this line and the qual score line
-            getline(instream, line);
-            // cout << "qual score line ." << line << "." << endl;
-        } else if (line.empty() or line[0] == '>' or line[0] == '@') {
-            return;
-        }
-    }
-}
-
-void print(std::ifstream& infile)
-{
-    char file;
-    std::vector<char> read;
-    uint i = 0;
-
-    // Read infile to vector
-    while (!infile.eof()) {
-        infile >> file;
-        read.push_back(file);
-    }
+    int read_status = kseq_read(this->inbuf);
 
-    // Print read vector
-    for (i = 0; i < read.size(); i++) {
-        std::cout << read[i];
+    bool no_more_reads_available = read_status == -1;
+    if (no_more_reads_available) {
+        return;
     }
-}
-
-void print(std::istream& infile)
-{
-    char file;
-    std::vector<char> read;
-    int i = 0;
 
-    // Read infile to vector
-    while (!infile.eof()) {
-        infile >> file;
-        read.push_back(file);
+    if (read_status == -2) {
+        throw "Truncated quality string detected";
+    } else if (read_status == -3) {
+        throw "Error reading " + this->filepath;
     }
 
-    // Print read vector
-    for (auto i = 0; i < read.size(); i++) {
-        std::cout << read[i];
-    }
+    ++this->num_reads_parsed;
+    this->name = this->inbuf->name.s;
+    this->read = this->inbuf->seq.s;
 }
 
 void FastaqHandler::get_id(const uint32_t& id)
 {
     const uint32_t one_based_id = id + 1;
-    if (one_based_id < num_reads_parsed) {
-        BOOST_LOG_TRIVIAL(warning)
-            << "restart buffer as have id " << num_reads_parsed << " and want id "
-            << one_based_id << " (" << id << ") with 0-based indexing.";
+    if (one_based_id < this->num_reads_parsed) {
         num_reads_parsed = 0;
         name.clear();
         read.clear();
-        line.clear();
-        assert(name.empty());
-        assert(read.empty());
-        assert(line.empty());
-
-        instream.ignore(std::numeric_limits<std::streamsize>::max());
-        fastaq_file.seekg(0, fastaq_file.beg);
-        instream.clear();
-        inbuf.pop();
-        if (gzipped) {
-            inbuf.pop();
-            inbuf.push(boost::iostreams::gzip_decompressor());
-        }
-        inbuf.push(fastaq_file);
+        gzrewind(this->fastaq_file);
+        kseq_rewind(this->inbuf);
     }
 
     while (id > 1 and num_reads_parsed < id) {
-        skip_next();
+        get_next();
         if (eof()) {
             break;
         }
     }
 
     while (num_reads_parsed <= id) {
         get_next();
-        if (eof()) {
-            break;
+        if (eof() && num_reads_parsed < id) {
+            throw std::out_of_range("Requested a read past the end of the file.");
         }
     }
 }
 
 void FastaqHandler::close()
 {
-    BOOST_LOG_TRIVIAL(debug) << "Close fastaq file";
-    fastaq_file.close();
+    if (!this->is_closed()) {
+        this->closed_status = gzclose(this->fastaq_file);
+    }
 }
```

Comment:
I might be being too pedantic here, if so, please ignore this comment.

I would prefer to change from `int closed_status` to `bool closed`. `int closed_status` implies that to understand the code, we have to know the return values of `gzclose()`. I also don't understand why `closed_status` is initialised with default 3. I tried to check what 3 means as a `closed_status` in the docs: https://refspecs.linuxbase.org/LSB_3.0.0/LSB-Core-generic/LSB-Core-generic/zlib-gzclose-1.html but I did not find any documentation in it. Also, we could have errors when closing this file, and we are not dealing with it (e.g. `gzclose()` can return `Z_STREAM_ERROR`, `Z_ERRNO` or `Z_BUF_ERROR`), and we would just say it is not closed.

I think it would be simpler and clearer for us in the future to maintain this code if `int closed_status` becomes `bool closed`, initialised with `false` by default. When closing the file, if the close was successful (i.e. `gzclose()` returned `Z_OK`), then `closed` becomes `true`. If `gzclose()` was unsuccessful, then an exception is thrown.

Reply:
You're right. I will rework this! I love code review!